### ISSUE TITLE
Improvement: Allow focus selector and switch windows to cross between screen while respecting they 2d layout.

### ISSUE THREAD

**Issue by andrei93r** [anametologin/krohnkite#64]
Hi,

I did not find an obvious way to switch between desktops however the ability to continue with the selector and switch the window would be greatly appreciated. I also might have an uncommon set-up (see below) since I have a dock to connect the laptop to 3 external displays, one of which is vertical. 

![Screenshot_20240831_160922](https://github.com/user-attachments/assets/5afcf622-7191-4240-97bf-e027ca346981)

Thank you for your amazing work,
Andrei

**Comment by Cal-Nine** [comment]
I use the basic KWin "Switch to Window to the Left/Right etc." instead of Krohnkite focus left/right/up/down as a workaround. Unfortunately moving windows between screens is still "next/previous" rather than directional.

**Comment by andrei93r** [comment]
@Cal-Nine 

Yes , I did find that eventually. Thank you for suggesting that. I believe my request is more ergonomic as one would not need one would not need to swtich to another key row just to move the window or focus on another screen . KDE itself has an option like this , but it only works with focus. Moving the windows would be a big plus

**Comment by this-username-system-sucks-ass** [comment]
I'd also like to see this if it's possible.

**Comment by UnsaltedScholar** [comment]
As a corollary to this -- maybe as a config option -- it would be nice if we could get something like COSMIC's ability to move windows between virtual desktops as well (i.e. move window+move focus to next desktop (direction dependent) and re-tile if needed). If I have time, I may try to implement such a thing myself, but I am unfamiliar with KDE/Krohnkite development. But aside from some logic to check if a virtual desktop exists in a certain direction, it seems fairly feasible. If this [KDE issue](https://discuss.kde.org/t/bug-fix-per-screen-virtual-desktops/16241/37) ever gets implemented, it would be even better.

**Comment by laurapigeon** [comment]
I would really appreciate this too- turning on `Multiscreen behaviour: Separate screen focus` as is suggested [here](https://github.com/anametologin/krohnkite?tab=readme-ov-file#setting-up-for-multi-screen) means that when changing virtual desktops, a window on the currently selected screen is focussed, rather than the last window selected on that desktop like usual, which throws me off. Having separate screen focus turned off removes the ability to use the `Switch to Next Screen` shortcut, and so without alt-tabbing or using the cursor I cannot focus between my screens. A toggle option to allow focus next/prev/left/right/up/down to pass across screen boundaries would be perfect. Even if its just that when a focus change key would not have an effect (due to being at the first/last/rightmost/bottommost/etc window on the screen) it looks for the first window on the next/prev/right/bottom/etc screen to switch focus to.

**Comment by Cal-Nine** [comment]
@UnsaltedScholar 
> it would be nice if we could get something like COSMIC's ability to move windows between virtual desktops as well

KWin already has shortcuts for this, but they are poorly labelled:
`Window One Desktop to the Left` `Window One Desktop Down` `Window to Desktop 3` etc.

If you do a search for "move window" they won't come up because they don't contain the word move like every other move window shortcut.

I'm not familiar with COSMIC but I believe this should do exactly what you need.


@laurapigeon 
> A toggle option to allow focus next/prev/left/right/up/down to pass across screen boundaries would be perfect

Try using KWin's `Switch to Window to the Left` etc. shortcuts instead of Krohnkite's shortcuts. I haven't tried but I would assume that would still work with separate screen focus turned off.

**Comment by UnsaltedScholar** [comment]
@Cal-Nine 

I am well aware of those KWin shortcuts. I am moreso referring to COSMIC's ability to shift workspaces dynamically with either the move focus or move window bindings. I cannot simultaneosly bind those kwin shortcuts with Krohnkite's own (I highly doubt doing so would provide the wanted functionality anyways). Perhaps, with additional logic similar to what I mentioned in my previous comment,one could call those kwin shortcuts inside of Krohnkite's binds to emulate COSMIC's functionality, but right now such is not implemented as far as I am aware.

**Comment by laurapigeon** [comment]
@Cal-Nine 
> Try using KWin's Switch to Window to the Left etc. shortcuts instead of Krohnkite's shortcuts. I haven't tried but I would assume that would still work with separate screen focus turned off.

That works thanks! I find the KWin built-in shortcuts slightly more imprecise than Krohnkite's, but the multiscreen behaviour is exactly what I wanted, and the last focussed window on each virtual desktop is remembered, regardless of screen.

**Comment by Cal-Nine** [comment]
> I find the KWin built-in shortcuts slightly more imprecise than Krohnkite's

From memory, they don't work in monocle mode – that is something I probably should have mentioned.
